### PR TITLE
Add some implementations of search algorithms in the trainer

### PR DESCRIPTION
**Pull request by xuanfeiren**
Upload **beamsearch_algorithm.py** and **UCBsearch.py** to opto/trainer/algorithms/.
In **beamsearch_algorithm.py**, I implemente two versions of beamsearch algorithms, standard one and one with history information provided in the feedback. In **UCBsearch.py**, I implement three versions of UCB search algorithms with a buffer.

Upload **example_usage_trainer.py** to examples. This example usage code could run different search algorithms on a simple math Q&A dataset, including MinibatchAlgorithm, BasicSearchAlgorithm and those I just uploaded.

### CODE REVIEW

**Comment by chinganc** [comment]
Start reviewing.

**Comment by xuanfeiren** [comment]
> @xuanfeiren Can you make the prompt templates used in the Trainer subclasses as attributes of the class? so that people can more easily modify it when needed without rewriting the entire function. Here is an example (https://github.com/xuanfeiren/Trace/blob/6b302c8fbe044a77f0091d228651ff2f64539dc5/opto/trainer/algorithms/UCBsearch.py#L537)

I have done that

File: opto/trainer/algorithms/UCBsearch.py [hunk 1]
```diff
@@ -0,0 +1,1009 @@
...
+                    # Fallback: if step modified agent in-place and didn't return dict, current agent state is a_prime
+                    a_prime_params_dict = {p: copy.deepcopy(p.data) for p in self.agent.parameters()}
+
+            except Exception as e:
+                print_color(f"Iter {iteration}: Error during optimizer.step for a_prime: {e}. Skipping candidate generation.", 'red')
+                continue
+            
+            # 4. Evaluate 'a_prime' on samples of validation set
+            a_prime_score, a_prime_evals = self._evaluate_candidate(
+                a_prime_params_dict, train_dataset, guide, evaluation_batch_size, num_threads # Use train_dataset and guide
```
[the remark this answers is not in the document]
Comment (xuanfeiren) [reply opto/trainer/algorithms/UCBsearch.py:247]: Yes there could be a validation set. I adopted your suggestion.
